### ISSUE TITLE
Design blog post template with reproducible experiment guarantee

### ISSUE THREAD

**Issue by niznik-dev** [niznik-dev/cruijff_kit#370]
## Summary

Explore options for publishing blog posts that showcase cruijff_kit experiments, and create a reusable template. Each post should come with a "chocolate bar guarantee" — a real, reproducible experiment that readers can run themselves.

## Goals

1. **Identify a publishing platform** — evaluate where these posts should live (group website, GitHub Pages, standalone blog, etc.)
2. **Create a blog post template** — a standardized structure stored in the wiki that authors can follow for future posts
3. **Define the reproducibility contract** — what exactly do we guarantee? That the experiment_summary.yaml produces the same experimental design? That results are qualitatively similar? Document this clearly in the template.

## Template should include

- Narrative structure (motivation, method, results, discussion)
- Pointer to the experiment_summary.yaml readers can use to reproduce
- Prerequisites and setup instructions
- Expected outputs and how to verify them

## Beta Readiness

This is item #2 on the [Beta Readiness](https://github.com/niznik-dev/cruijff_kit/wiki/Beta-Readiness) checklist (New Capabilities).

**Comment by niznik-dev** [comment]
I've been running some small sanity-check experiments on the side that could serve as the "chocolate bar" for the first blog post using this template. Since the choice of featured task shapes what the template needs to support (length, narrative structure, what counts as "reproduced"), sharing a quick comparison here.

**Three completed candidates (all Llama-3.2-3B-Instruct, LoRA r8):**

| Task | In-dist | OOD | Calibration signal | What's interesting |
|------|---------|-----|--------------------|-------------------|
| **Bernoulli** (p=0.7 vs 0.3, len 10) | 93.4% (AUC 0.977) | — | Model P(A) correlates r=0.992 with the Bayesian posterior, RMSE 0.055 | Clean pipeline validation against a known ground truth |
| **Count digits** (train len 3–6, eval len 3–10) | 100% | 1B: 0% at every OOD length. 3B: 78/60/46/84% at len 7/8/9/10 | — | Hard ceiling at the training-max answer; partial breakthrough at len 10 (two-char token); possible string-length shortcut confound |
| **Majority** (train len 5–15, eval len 5–25) | 99.4% | 94.8% best, decays with training | AUC ~1.0 both splits; OOD ECE grows 0.022 → 0.031 → 0.039 across epochs | In-dist accuracy climbs 98.6 → 99.0 → 99.4 while OOD *peaks at epoch 0* and degrades with further training |

**Proposal: lead the first post with majority.** Three reasons:

1. **It has an actual narrative arc.** Simple task to explain, strong headline number, and a specific teachable finding — OOD performance peaks early and silently degrades even as in-distribution accuracy improves. That's a real lesson about held-out evaluation, not just a "pipeline works" green light.
2. **It exercises the reproducibility contract more stringently.** Bernoulli is lovely but a reader running it just recovers "yes, the calibration stack works." Majority has a specific shape to the OOD curve across epochs that a reader can verify — which is a meaningful test of what "reproduced" should mean in the contract.
3. **Count-digits has the most interesting raw result, but the story isn't stable enough yet.** Non-monotonic OOD (78/60/46/84) is striking, but there's a plausible string-length shortcut confound that needs an explicit ablation before we'd want to lead with it.

Supporting roles if we go this way: Bernoulli as a short "how we check calibration" sidebar, count-digits as a "here's an open question we're chasing" teaser.

**Before publishing the first post, I'd want:**
- [ ] Seed stability on the epoch-0 OOD peak (is 94.8% a single-seed artifact?)
- [ ] Error analysis by margin (do OOD failures concentrate on narrow-margin sequences?)
- [ ] Decision on whether to add a 1B comparison run — would strengthen the post but not a blocker

—MxC

**Comment by niznik-dev** [comment]
**Phase 1 and Phase 1b results update.** Following our [earlier comment](https://github.com/niznik-dev/cruijff_kit/issues/370#issuecomment-4224774748) proposing majority as the featured task, we ran the seed stability check and a follow-up schedule test. The original pitch relied on one trajectory detail that didn't replicate, but the revised story (still lead with majority) is cleaner and sits on stronger evidence.

## Follow-up results

**Phase 1 (5 seeds × 5 epochs):** all 5 seeds showed OOD accuracy *improving* with training, not the "peaks at epoch 0 and degrades" pattern the 2026-04-09 run suggested. 0 of 5 replicated the pattern.

**Phase 1b (2×2 schedule vs seed):** to separate schedule effects (189-step 3-epoch run vs 315-step 5-epoch run) from seed effects, ran seed 1234 at 5 epochs and 3 new seeds at 3 epochs. Result:

| | seed 1234 | seeds 7/23/47 |
|---|---|---|
| **3 epochs** | degrades (original baseline) | 3/3 improve |
| **5 epochs** | improves | 3/3 improve |

One cell out of four shows OOD degradation: the original baseline run. Every other seed × schedule combination improves or stays flat. The epoch-0 peak is a single-seed result — not a property of seed 1234, not a property of the 3-epoch schedule.

## The finding that holds up

The better frame for a sanity check is "does the pipeline produce a model that learns the task?" — answered in three parts:

| Question | Evidence | Answer |
|---|---|---|
| Does the model fit the training distribution? | Training-set accuracy: **99.90% for the best 5-epoch run (seed 23 at epoch 4, 2 misses out of 2000), 99.15% for the original baseline (17 misses out of 2000)**. AUC = 1.000 on both — the model's confidence ordering is perfect even where its argmax is wrong. | Approximately interpolates, modulo hard cases |
| Does it generalize within distribution? | Held-out in-dist (lengths 5–15) accuracy: **97–100% across all 10 runs** | Yes |
| Does it generalize *out* of distribution? | Held-out OOD (lengths 16–25) accuracy: **93.4–99.0% across all 10 runs, mean ≈ 96.5%** | **Yes, reproducibly** |

Ten independent runs (varying seed and schedule) all achieve ≥93% OOD on sequences up to 67% longer than anything seen during training. Fine-tuning teaches the model a generalizing function on this task, not a lookup table — the residual training errors concentrate on narrow-margin sequences where the task is inherently hardest, and generalization to OOD samples tracks training fit cleanly. The best OOD generalizer (seed 23, 99.0% OOD) is also the best training-set fitter (99.90%) — fit and generalization go together, no tradeoff.

## Contrast with count-digits

Another cruijff_kit sanity check, structurally similar (count a property of a short numeric sequence):

| Task | In-dist | OOD (longer sequences) |
|---|---|---|
| Count digits, 1B r64 | 100% | **0%** — hard ceiling at training-max answer |
| Count digits, 3B r8 | 100% | 46–84%, non-monotonic across lengths |
| **Majority, 3B r8 (10 runs)** | **99–100%** | **93–99%** |

Same fine-tuning pipeline, same hyperparameters, same dataset size. Majority has a binary output space; count-digits has an open-vocabulary numeric output. One generalizes OOD, the other doesn't. That contrast is the pedagogical payload of the post.

## Proposal

Still lead with majority — for this reason:

1. **Setup:** given a binary sequence, identify the majority digit.
2. **Training:** fine-tune a 3B r8 model on 2000 sequences of length 5–15 (≈5 minutes on one A100).
3. **Check 1 — fit:** 99–100% training accuracy (approximately interpolates), 99–100% on held-out in-dist.
4. **Check 2 — generalization:** 93–99% on held-out OOD sequences of length 16–25. Yes, reproducibly across seeds.
5. **Contrast:** show count-digits' failure mode. Same pipeline, different output space, totally different OOD behavior. Unpack why.
6. **Chocolate bar guarantee:** readers running with their own seed should see training fit ≥ 99%, in-dist ≥ 97%, and OOD in the 93–99% range.

## Methodology note

Cross-seed OOD variance at intermediate training depths can be large. In the 5-epoch runs, the OOD spread across seeds at epoch 2 was **10.2 percentage points** (min 89.4%, max 99.6%). At epoch 4 it shrinks to 4.2pp. If the template specifies a reproducibility contract, it should probably use final-epoch ranges (e.g., "OOD accuracy 93–99%") rather than full-trajectory matching.

—MxC

**Comment by msalganik** [comment]
@niznik-dev I propose that we move this out of 0.3 since this is about attracting new users and I think that is not our priority right now. We can come back to this when the timing is better.

**Comment by niznik-dev** [comment]
<img src="https://raw.githubusercontent.com/niznik-dev/llm-personas/main/fusions/mattie/mattie-icon.png" width="36" align="absmiddle"> 🎻 **MxC** ✳️

Closing as part of the project archive. No further blog posts are planned for cruijff_kit, so the platform and template questions are moot.

—MxC
